Analysis/bin/ShuffleMergeSpectral.cxx

```diff
@@ -0,0 +1,681 @@
+            throw analysis::exception("Invalid source weight for the source '%1%'.") % name;
+        if(!files_n_total)
+            throw analysis::exception("Empty source '%1%'.") % name;
+
+    }
+
+    SourceDesc(const SourceDesc&) = delete;
+    SourceDesc& operator=(const SourceDesc&) = delete;
+
+    bool DoNextStep()
+    {
+
+      if(current_file_index == files_n_total-1 && current_n_processed >= entries_file) return false;
+
+      while(!current_file_index || current_n_processed == entries_file) {
+          if(!current_file_index)
+              current_file_index = 0;
+          else
+              ++(*current_file_index);
+          if(*current_file_index >= file_names.size())
+              throw analysis::exception("The expected number of events = %1% is bigger than the actual number of"
+                                        " events in source '%2%'.") % entries_file % name;
+          current_n_processed = 0;
+          const std::string& file_name = file_names.at(*current_file_index);
+          current_tuple.reset();
+          current_file = root_ext::OpenRootFile(file_name);
+          current_tuple = std::make_shared<TauTuple>("taus", current_file.get(), true, disabled_branches);
+          entries_file = current_tuple->GetEntries();
+          if(!entries_file)
+            throw analysis::exception("Root file inside %1% is empty.") % name;
+      }
+      ++total_n_processed;
+      current_tuple->GetEntry(current_n_processed++);
+      (*current_tuple)().dataset_id = std::hash<std::string>{}(name);
+      (*current_tuple)().dataset_group_id = std::hash<std::string>{}(group_name);
+      return true;
+    }
+
+    const Tau& GetNextTau()
+    {
+      return current_tuple->data();
+    }
+
+    size_t GetNumberOfEvents() const { return entries_file; }
+    const std::string& GetName() const { return name; }
+    const std::vector<std::string>& GetFileNames() const { return file_names; }
+    double GetWeight() const { return weight; }
+    // const std::string& GetBinName() const { return bin_name; }
+    // void SetBinName(const std::string& _bin_name) { bin_name = _bin_name; }
+
+  private:
+      const std::string name;
+      const std::string group_name;
+      const std::vector<std::string> file_names;
+      const std::set<std::string> disabled_branches;
+      const double weight;
+      std::shared_ptr<TFile> current_file;
+      std::shared_ptr<TauTuple> current_tuple;
+      boost::optional<ULong64_t> current_file_index;
+
+      // types should be checked
+      ULong64_t current_n_processed;
+      ULong64_t files_n_total;;
+      size_t entries_file;
+      size_t total_n_processed;
+  };
+
+
+struct EntryDesc {
+    using TauType = analysis::TauType;
+    using SampleType = analysis::SampleType;
+
+    std::string name;
+    std::map< std::string, std::vector<std::string>> data_files;
+    std::map< std::string, std::string> spectrum_files;
+    std::vector< std::string > tau_types;
```

Comment:
- better to define as `std::set<TauType> tau_types;`
- everywhere in the code, to represent tau types use `TauType` instead of `std::string`
https://github.com/cms-tau-pog/TauMLTools/blob/master/Analysis/interface/AnalysisTypes.h#L53